### ISSUE TITLE
Incorrect `dead_code` lint with generics using type aliases in macros.

### ISSUE THREAD

**Issue by zachs18** [rust-lang/rust#118424]
### Code

```Rust
macro_rules! maybe_generic_static_assert {
    (generic = $T:ident, $expr:expr, $msg:expr) => {
        mod _generic_helper {
            type $T = u32;
            const _: () = assert!($expr, $msg);
        }
    };
    ($expr:expr, $msg:expr) => {
        const _: () = assert!($expr, $msg);
    }
}

maybe_generic_static_assert!(
    std::mem::align_of::<u32>() == std::mem::align_of::<std::sync::atomic::AtomicU32>(),
    "aaa"
);

maybe_generic_static_assert!(
    generic = T,
    std::mem::align_of::<*mut T>() == std::mem::align_of::<std::sync::atomic::AtomicPtr<T>>(),
    "aaa"
);
```

### Current output

```Shell
Compiling playground v0.0.1 (/playground)
warning: type alias `T` is never used
  --> src/lib.rs:19:15
   |
19 |     generic = T,
   |               ^
   |
   = note: `#[warn(dead_code)]` on by default
   = note: this warning originates in the macro `maybe_generic_static_assert` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: `playground` (lib) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 2.11s
```

### Desired output

```Shell
<no warnings>
```

### Rationale and extra context

([Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a7e4539c37a7e7eecf42c712bfe4c50d))

The type alias `type $T = u32;` is not unused; and removing it causes the code to fail to compile.

The second invocation of the macro is the important one here; the `$expr:expr` references the type alias created from the passed `$T:ident`, so the type alias is not unused.

### Anything else?

Output with `-Zmacro-backtrace`

```rs
warning: type alias `T` is never used
  --> <source>:19:15
   |
1  | / macro_rules! maybe_generic_static_assert {
2  | |     (generic = $T:ident, $expr:expr, $msg:expr) => {
3  | |         mod _generic_helper {
4  | |             type $T = u32;
...  |
10 | |     }
11 | | }
   | |_- in this expansion of `maybe_generic_static_assert!`
...
18 | / maybe_generic_static_assert!(
19 | |     generic = T,
   | |               ^
20 | |     std::mem::align_of::<*mut T>() == std::mem::align_of::<std::sync::atomic::AtomicPtr<T>>(),
21 | |     "aaa"
22 | | );
   | |_- in this macro invocation
   |
   = note: `#[warn(dead_code)]` on by default

warning: 1 warning emitted
```

**Comment by zachs18** [comment]
@rustbot label -A-macros

Upon further minimizing, this actually appears to have nothing to do with the type alias being declared in a macro; the warning shows up even for 
```rs
type T = u32;
const _: () = assert!(
    std::mem::size_of::<T>() == std::mem::size_of::<T>()
);
const _A: () = assert!(
    std::mem::size_of::<T>() == std::mem::size_of::<T>()
);
```

The `const _A: () = { ... };` should definitely be quieting the lint for anything used in `...`, and I *think* the `const _: () = { ... };` should too, but neither do.

**Comment by long-long-float** [comment]
The problem happens without `assert!`.

```rust
type T = u32;
const _A: usize = std::mem::size_of::<T>();

fn main() {}
```

**Comment by zachs18** [comment]
This also happens with several other kinds of items, not just type aliases:
```Rust
type T = u32;
const fn hmm() -> usize { 0 }
struct Struct;
const B: usize = 42;
impl Struct {
    const fn hmmm(&self) -> usize { 0 }
}

const _A: usize = 
    std::mem::size_of::<T>()
    + hmm()
    + std::mem::size_of::<Struct>()
    + B
    + Struct.hmmm();
fn main() {
  // let _a = _A; // uncomment for no warnings
}
```

```Rust
   Compiling playground v0.0.1 (/playground)
warning: type alias `T` is never used
 --> src/main.rs:1:6
  |
1 | type T = u32;
  |      ^
  |
  = note: `#[warn(dead_code)]` on by default

warning: function `hmm` is never used
 --> src/main.rs:2:10
  |
2 | const fn hmm() -> usize { 0 }
  |          ^^^

warning: struct `Struct` is never constructed
 --> src/main.rs:3:8
  |
3 | struct Struct;
  |        ^^^^^^

warning: constant `B` is never used
 --> src/main.rs:4:7
  |
4 | const B: usize = 42;
  |       ^

warning: method `hmmm` is never used
 --> src/main.rs:6:14
  |
5 | impl Struct {
  | ----------- method in this implementation
6 |     const fn hmmm(&self) -> usize { 0 }
  |              ^^^^

warning: `playground` (bin "playground") generated 5 warnings
    Finished dev [unoptimized + debuginfo] target(s) in 2.55s
     Running `target/debug/playground`
```

([Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b544e366dbcc71c81b4f0bc88eecc431))
